### PR TITLE
Fix and unify markdown usage in 'Analyzer Configuration.md'

### LINKED ISSUE #4516: Properly escape special characters in Analyzer Configuration.md
Parts of this markdown file doesn't currently render correctly.

e.g.:

image

image

### PR DESCRIPTION
**Pull request by Evangelink**
Fix #4516

### CODE REVIEW

File: docs/Analyzer Configuration.md [hunk 1]
```diff
@@ -1,10 +1,12 @@
+<!-- Markdown used in this file is tweaked for GitHub rendering, it's possible that it renders differently in other tools (e.g. VS Code) -->
```
Comment (Youssef1313) [inline docs/Analyzer Configuration.md:1]: If that happens, then we're violating some markdown rules.
It *should* render the same.

Comment (Evangelink) [reply docs/Analyzer Configuration.md:1]: Nope, each rendering engine is doing a different "interpretation" of markdown.

Comment (Evangelink) [reply docs/Analyzer Configuration.md:1]: That's why you have such extensions: https://marketplace.visualstudio.com/items?itemName=bierner.markdown-preview-github-styles

File: docs/Analyzer Configuration.md [hunk 2]
```diff
@@ -229,29 +231,29 @@ Configurable Rules: [CA2241](https://docs.microsoft.com/dotnet/fundamentals/code
...
 Examples:
 
 | Option Value | Summary |
 | --- | --- |
-|`dotnet_code_quality.additional_string_formatting_methods = MyFormat` | Matches all methods named 'MyFormat' in the compilation
-|`dotnet_code_quality.additional_string_formatting_methods = MyFormat1\|MyFormat2` | Matches all methods named either 'MyFormat1' or 'MyFormat2' in the compilation
-|`dotnet_code_quality.additional_string_formatting_methods = NS.MyType.MyFormat(ParamType)` | Matches specific method 'MyFormat' with given fully qualified signature
-|`dotnet_code_quality.additional_string_formatting_methods = NS1.MyType1.MyFormat1(ParamType)\|NS2.MyType2.MyFormat2(ParamType)` | Matches specific methods 'MyFormat1' and 'MyFormat2' with respective fully qualified signature
+| `dotnet_code_quality.additional_string_formatting_methods = MyFormat` | Matches all methods named 'MyFormat' in the compilation. |
+| `dotnet_code_quality.additional_string_formatting_methods = MyFormat1\|MyFormat2` | Matches all methods named either 'MyFormat1' or 'MyFormat2' in the compilation. |
```
Comment (Youssef1313) [inline docs/Analyzer Configuration.md:245]: Is this one of the places where it doesn't render correctly in VSCode?

File: docs/Analyzer Configuration.md [hunk 3]
```diff
@@ -654,17 +656,17 @@ Configurable Rules: [CA1010](https://docs.microsoft.com/dotnet/fundamentals/code
...
-Default Value: None
+Default Value: _None_
 
 Examples:
 
 | Option Value | Summary |
 | --- | --- |
-|`dotnet_code_quality.CA1010.additional_required_generic_interfaces = ISomething->System.Collections.Generic.IEnumerable`\``1` | All types implementing 'ISomething' regardless of its namespace are expected to also implement 'System.Collections.Generic.IEnumerable`1'. |
-|`dotnet_code_quality.CA1010.additional_required_generic_interfaces = T:System.Collections.IDictionary->T:System.Collections.Generic.IDictionary`\``2` | All types implementing 'System.Collections.Generic.IDictionary' are expected to also implement 'System.Collections.Generic.IDictionary`2'. |
+| ``dotnet_code_quality.CA1010.additional_required_generic_interfaces = ISomething->System.Collections.Generic.IEnumerable`1`` | All types implementing 'ISomething' regardless of its namespace are expected to also implement 'System.Collections.Generic.IEnumerable\`1'. |
```
Comment (Youssef1313) [inline docs/Analyzer Configuration.md:668]:
```suggestion
| ``dotnet_code_quality.CA1010.additional_required_generic_interfaces = ISomething->System.Collections.Generic.IEnumerable`1`` | All types implementing 'ISomething' regardless of its namespace are expected to also implement ``System.Collections.Generic.IEnumerable\`1``. |
```

File: docs/Analyzer Configuration.md (hunk 2 above)
Comment (Evangelink) [reply docs/Analyzer Configuration.md:245]: Yes, for GitHub it is rendered as `|` while it is rendered as `\|` for VS Code.

File: docs/Analyzer Configuration.md [hunk 4]
```diff
@@ -654,17 +656,17 @@ Configurable Rules: [CA1010](https://docs.microsoft.com/dotnet/fundamentals/code
...
+| ``dotnet_code_quality.CA1010.additional_required_generic_interfaces = T:System.Collections.IDictionary->T:System.Collections.Generic.IDictionary`2`` | All types implementing 'System.Collections.Generic.IDictionary' are expected to also implement 'System.Collections.Generic.IDictionary`2'. |
```
Comment (Youssef1313) [inline docs/Analyzer Configuration.md:669]:
```suggestion
| ``dotnet_code_quality.CA1010.additional_required_generic_interfaces = T:System.Collections.IDictionary->T:System.Collections.Generic.IDictionary`2`` | All types implementing 'System.Collections.Generic.IDictionary' are expected to also implement ``System.Collections.Generic.IDictionary`2``. |
```

File: docs/Analyzer Configuration.md (hunk 3 above)
Comment (Evangelink) [reply docs/Analyzer Configuration.md:668]: Another example where it is shown correctly in GitHub but not in VS Code.

File: docs/Analyzer Configuration.md (hunk 2 above)
Comment (Youssef1313) [reply docs/Analyzer Configuration.md:245]: Can you try something like the following and check if it renders correctly in both?

```md
<code>dotnet_code_quality.additional_string_formatting_methods = NS1.MyType1.MyFormat1(ParamType)&#124;NS2.MyType2.MyFormat2(ParamType)</code>
```

File: docs/Analyzer Configuration.md (hunk 1 above)
Comment (Youssef1313) [reply docs/Analyzer Configuration.md:1]: I think there's always a way to tweak things to render correctly in both.

Comment (Youssef1313) [reply docs/Analyzer Configuration.md:1]: For example, see the rationale written for MD022

https://github.com/DavidAnson/markdownlint/blob/main/doc/Rules.md#md022---headings-should-be-surrounded-by-blank-lines

> Rationale: Aside from aesthetic reasons, some parsers, including kramdown, will not parse headings that don't have a blank line before, and will parse them as regular text.

Comment (Evangelink) [reply docs/Analyzer Configuration.md:1]: If you find a way to have things to render correctly please I am open to suggestion :)

File: docs/Analyzer Configuration.md (hunk 4 above)
Comment (Youssef1313) [reply docs/Analyzer Configuration.md:669]: This one is expected to not render correctly before the fix. The backtick wasn't escaped, nor included between double backticks.

File: docs/Analyzer Configuration.md (hunk 1 above)
Comment (Youssef1313) [reply docs/Analyzer Configuration.md:1]: I haven't tried, but the suggestion in https://github.com/dotnet/roslyn-analyzers/pull/4541#discussion_r539356934 may work.

IIRC, I've seen this in dotnet/docs

File: docs/Analyzer Configuration.md (hunk 4 above)
Comment (Evangelink) [reply docs/Analyzer Configuration.md:669]: We have to decide if for the description we want to keep single quote (in which case the right fix is "\`") or if we want to go for "code" in which case your solution is good.

File: docs/Analyzer Configuration.md (hunk 2 above)
Comment (Evangelink) [reply docs/Analyzer Configuration.md:245]: MD0033 no inline html code :)

File: docs/Analyzer Configuration.md (hunk 4 above)
Comment (Youssef1313) [reply docs/Analyzer Configuration.md:669]: The docs team always goes for code for any code element name.

btw, I think there is a typo in this sentence, there is an extra "Generic" in the namespace for non-generic `IDictionary` interface:

> All types implementing 'System.Collections.~~Generic.~~IDictionary'

@Evangelink Can you confirm?

Comment (Evangelink) [reply docs/Analyzer Configuration.md]: Namespace does include `Generic`, see https://docs.microsoft.com/dotnet/api/system.collections.generic.idictionary-2

File: docs/Analyzer Configuration.md (hunk 1 above)
Comment (Evangelink) [reply docs/Analyzer Configuration.md:1]: MD0033 no inline html code :)

Comment (Youssef1313) [reply docs/Analyzer Configuration.md:1]: It's a choice between having consistency over different parsers vs. violating MD033. Probably we should have a feature request for markdownlint to detect such cases and exclude them for MD033 :smile:

File: docs/Analyzer Configuration.md (hunk 4 above)
Comment (Youssef1313) [reply docs/Analyzer Configuration.md]: @Evangelink There are two `IDictionary` interfaces being referred to here. The generic one which has a `2 at the end, and the non-generic one:

https://docs.microsoft.com/en-us/dotnet/api/system.collections.idictionary?view=net-5.0

Comment (Evangelink) [reply docs/Analyzer Configuration.md]: Oh sorry I was looking only at the second part!

**Review by mavasani — APPROVED** [review]
LGTM

@Youssef1313 can you please also confirm this will not conflict with your PR to enable markdown validation for the repo?

**Comment by Youssef1313** [comment]
> @Youssef1313 can you please also confirm this will not conflict with your PR to enable markdown validation for the repo?

It should be okay.

**Pull request merged**